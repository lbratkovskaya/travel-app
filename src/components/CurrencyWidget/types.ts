export type CurrencyData = {
  rates:
  { [key: string]: { rate: number } },
};
