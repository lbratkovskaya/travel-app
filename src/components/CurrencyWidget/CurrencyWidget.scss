@import '../widgets.scss';

.currency-widget {
  @include widget;

  .currency-widget-icon {
    color: $icon-color;
  }

  .currency-widget-data {
    display: flex;
    flex-direction: column;

    .curr-key {
      font-size: 1rem;
    }
  }
}
