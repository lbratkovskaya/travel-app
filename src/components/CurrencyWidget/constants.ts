const API_URL_NO_PARAM = 'https://rs-travel-app.herokuapp.com/currencies?';
// const API_URL_NO_PARAM = 'http://localhost:8000/currencies?';

const ROUND_RATE = 10000;
export {
  API_URL_NO_PARAM,
  ROUND_RATE,
};
