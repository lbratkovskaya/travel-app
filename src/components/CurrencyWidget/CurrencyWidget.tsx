import React, { useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';
import { useTranslation } from 'react-i18next';
import { CircularProgress } from '@material-ui/core';
import { AttachMoneyOutlined } from '@material-ui/icons';
import rootConnector,
{
  rootProps,
} from '../../store/rootConnector';
import { Country, URLParamTypes } from '../../types';
import { CurrencyData } from './types';
import {
  API_URL_NO_PARAM,
  ROUND_RATE,
} from './constants';
import './CurrencyWidget.scss';

const CurrencyWidget: React.FC<rootProps> = (props: rootProps) => {
  const { countryId } = useParams<URLParamTypes>();
  const [currencyData, setCurrencyData] = useState<{ [key: string]: number } | null>({});
  const [isLoading, setLoading] = useState<boolean>(true);
  const { countries } = props;

  const { t } = useTranslation();

  const currentCountry: Country | undefined = countries && countries
    .find((country: Country) => country.id === countryId);
  const currency: string | undefined = currentCountry?.currency;

  useEffect(() => {
    if (currency) {
      const API_URL = `${API_URL_NO_PARAM}curr=${currency}`;

      fetch(API_URL)
        .then((response) => {
          if (!response.ok) throw new Error(response.statusText);
          return response;
        })
        .then((res) => res.json())
        .then((data: CurrencyData) => {
          const fetchedData: { [key: string]: { rate: number } } = { ...data.rates };
          const cData = {} as { [key: string]: number };

          cData.USD = 1 / fetchedData[`USD${currency}`].rate;
          cData.EUR = cData.USD * fetchedData.USDEUR.rate;
          cData.RUB = cData.USD * fetchedData.USDRUB.rate;
          setCurrencyData(cData);
          setLoading(false);
        })
        .catch(() => {
          setCurrencyData(null);
          setLoading(false);
        });
    }
  }, []);

  const getCurrencies = (): JSX.Element[] => Object.entries(currencyData || {})
    .map(([key, value]) => (
      <div key={key}>
        <span>{1}</span>
        <span className="curr-key">{`${currency} = `}</span>
        <span>{`${Math.round(value * ROUND_RATE) / ROUND_RATE} `}</span>
        <span className="curr-key">{`${key}`}</span>
      </div>
    ));

  const countryName: string = t(`${currentCountry?.id}.name`);
  const errorMessage: string = t('currency_fetch_error', { countryName });
  return (
    <div className="currency-widget">
      <div className="currency-widget-icon">
        <AttachMoneyOutlined />
      </div>
      {isLoading ? <CircularProgress /> : null}
      {currencyData
        ? (
          <div className="currency-widget-data">
            {getCurrencies()}
          </div>
        )
        : (
          <div className="widget-error">
            {errorMessage}
          </div>
        )}
    </div>
  );
};

export default rootConnector(CurrencyWidget);
