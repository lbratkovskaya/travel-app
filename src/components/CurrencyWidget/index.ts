import CurrencyWidget from './CurrencyWidget';

export default CurrencyWidget;
