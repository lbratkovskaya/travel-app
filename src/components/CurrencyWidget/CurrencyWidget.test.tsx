import React from 'react';
import { Provider } from 'react-redux';
import { Route, Router } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { act } from 'react-dom/test-utils';
import { createMemoryHistory } from 'history';
import { mount, shallow, ShallowWrapper } from 'enzyme';
import CurrencyWidget from './CurrencyWidget';

const mockStore = configureStore([]);

const runAllPromises = () => new Promise(setImmediate);

describe('CurrencyWidget functional component testing', () => {
  let wrapper: ShallowWrapper<typeof CurrencyWidget>;

  const initialState = {
    countries: [
      {
        id: 'test',
        currency: 'RUB',
      },
    ],
  };
  const store = mockStore(initialState);

  beforeEach(() => {
    wrapper = shallow<typeof CurrencyWidget>(<Provider store={store}><CurrencyWidget /></Provider>);
  });

  it('CurrencyWidget must have a correct snapshot', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  const data = { rates: { USDEUR: { rate: 1 }, USDRUB: { rate: 3 } } };

  global.fetch = jest.fn(() => Promise.resolve({
    json: () => Promise.resolve(data),
    ok: true,
  } as Response));

  it('CurrencyWidget must have 3 rows', async () => {
    const history = createMemoryHistory({
      initialEntries: ['/country/test'],
    });

    const component = mount(
      <Provider store={store}>
        <Router history={history}>
          <Route path="/country/:countryId">
            <CurrencyWidget />
          </Route>
        </Router>
      </Provider>,
    );

    await act(async () => {
      await runAllPromises();
      component.update();
      expect(component.find('.currency-widget-data').children()).toHaveLength(3);
    });
  });
});
