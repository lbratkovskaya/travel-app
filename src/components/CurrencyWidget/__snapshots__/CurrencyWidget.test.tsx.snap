// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyWidget functional component testing CurrencyWidget must have a correct snapshot 1`] = `
"<ContextProvider value={{...}}>
  <Connect(CurrencyWidget) />
</ContextProvider>"
`;
